aptos-move/aptos-vm-logging/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+// as speculative logging event (and eventual dispatching as needed).
+impl SpeculativeEvent for VMLogEntry {
+    fn dispatch(self) {
+        match self.level {
+            Level::Error => {
+                // TODO: Consider using SpeculativeCounter to increase CRITICAL_ERRORS
+                // on the critical path instead of async dispatching.
+                error!(self.context, "{}", self.message);
+                CRITICAL_ERRORS.inc();
+            },
+            Level::Warn => warn!(self.context, "{}", self.message),
+            Level::Info => info!(self.context, "{}", self.message),
+            Level::Debug => debug!(self.context, "{}", self.message),
+            Level::Trace => trace!(self.context, "{}", self.message),
+        }
+    }
+}
+
+static BUFFERED_LOG_EVENTS: Lazy<ArcSwapOption<SpeculativeEvents<VMLogEntry>>> =
+    Lazy::new(|| ArcSwapOption::from(None));
+
+/// Initializes the storage of speculative logs for num_txns many transactions.
+pub fn init_speculative_logs(num_txns: usize) {
+    BUFFERED_LOG_EVENTS.swap(Some(Arc::new(SpeculativeEvents::new(num_txns))));
+}
+
+/// Adds a message at a specified logging level and given context (that includes txn index)
+/// to speculative buffer. Logs directly and logs a separate (new error) if the speculative
+/// events storage is not initialized or appropriately sized.
+pub fn speculative_log(level: Level, context: &AdapterLogSchema, message: String) {
+    let txn_idx = context.txn_idx;
+    match &*BUFFERED_LOG_EVENTS.load() {
+        Some(log_events) => {
+            let log_event = VMLogEntry::new(level, context.clone(), message);
+            if let Err(e) = log_events.record(txn_idx, log_event) {
+                error!("{:?}", e);
+                CRITICAL_ERRORS.inc();
+            };
+        },
+        None => {
+            error!(
+                "Speculative state not initialized to log message = {}",
+                message
+            );
+            CRITICAL_ERRORS.inc();
+        },
+    };
+}
+
+/// Flushes the currently stored logs, and swaps the speculative log / event storage with None.
+/// Must be called after block execution is complete (removes the storage from Arc).
+pub fn flush_speculative_logs() {
+    match BUFFERED_LOG_EVENTS.swap(None) {
+        Some(log_events_ptr) => match Arc::try_unwrap(log_events_ptr) {
+            Ok(log_events) => log_events.flush(),
+            Err(_) => {
+                error!("Speculative log storage must be uniquely owned to flush");
+                CRITICAL_ERRORS.inc();
+            },
+        },
+        None => {
+            error!("Flush called on uninitialized speculative log storage");
+            CRITICAL_ERRORS.inc();
+        },
+    }
+}
+
+/// Clear recorded speculative logs for all transactions, useful if the whole block
+/// execution is being ignored (e.g. currently, the module r/w fallback scenario).
+pub fn clear_all_speculative_logs() {
+    match &*BUFFERED_LOG_EVENTS.load() {
```

Comment:
it seems easier to just swap with a None?

Reply:
sure, deleted this whole function and re-initialized (can't just swap with None) from the caller side.